Add `webfinger-domain` configuration

tesaguri opened the pull request:
This adds ~~`instance.webfinger_domain`~~ `instance.webfinger-domain` configuration option to set a custom domain name for the `subject` of local WebFinger resources. See #285 for the background.

The configuration will look like the following:

```toml
[instance]
# ...
webfinger-domain = "example.com"

[url]
scheme = "https"
domain = "social.example.com"
```

This example will run the server at `social.example.com` and make the WebFinger handler resolve local actors as `acct:{username}@example.com` instead of `acct:{username}@social.example.com`.

`GET /api/v1/instance` of Mastodon API and its counterpart in GraphQL API is updated to use the new configuration value to make it align with Mastodon's implementation (though Mastodon doesn't provide a GraphQL API. But since the old value can trivially be known from the endpoint URL, I think the change makes sense for the GraphQL API as well).

## The name?

The configuration name is up for bikeshedding. There are few alternatives in my mind:

- `local-domain` (consistent with Mastodon, arguably the most popular implementaion. Might be confusing though)
- `webfinger-host`(might be more accurate since it may contain a port in addition to the domain, but is inconsistent with the existing `url.domain` config)
- `acct-domain` or `acct-host` (in reference to the `acct` field of Mastodon API)
- `display-domain` (adopted from [Takahē][takahe-doc-domains]. Personally I like it too because of consistency with the term "display name" for `@user` handle name)
- `account-domain` (adopted from [GoToSocial][gts-doc-split-domain])

[takahe-doc-domains]: <https://docs.jointakahe.org/en/stable/domains/>
[gts-doc-split-domain]: <https://docs.gotosocial.org/en/latest/advanced/host-account-domain/#configuration>

I chose the name `webfinger-domain` because it is consitent with `url.domain` and also helpfully reminds the administrator that it's about WebFinger, for which they need to set up an HTTP redirection at the pointed domain.

aumetra commented:
Sorry about the merge that messed things up. You should probably be able to pretty easily rebase though.  
It was just me moving around directories.

I'm gonna have a look at it after work.

aumetra commented:
Also just as a small pre-nit, can you rename the configuration option to be written in kebab case to be in line with the other options?

`webfinger_domain` -> `webfinger-domain`

tesaguri commented:
Rebased onto master.

> Edit: Never mind, should be handled already by the `serde` annotation

Yeah, I've also forgotten about `#[serde(rename_all = "kebab-case")]` and wrote a wrong description. Updated it as well.

aumetra commented:
How does our Webfinger resolver interact with these display domains? It doesn't respect the display domain IIRC, so that should probably be patched, too.  
The `domain` field in our database, that we use to distinguish servers, is filled by parsing the actor URL (which will pretty much always be to the canonical domain; i.e. `social.example.org` and not `example.org`).

In general we will have a harder time of supporting this completely since Mastodon *really* likes the Webfinger protocol and pretty much inseparably joins ActivityPub and Webfinger.  
It will fetch a user by its URL, use the username from the actor to send a Webfinger request, to obtain the display domain.

Kitsune doesn't do this. We see Webfinger as a QoL improvement that is stitched onto ActivityPub. Kitsune works fine without Webfinger.

I'm not entirely sure how it will interact with all other parts if we just patch an option into `FetchOpts` for the `Fetcher` structure.

tesaguri commented:
> so that should probably be patched, too.

Do you mean that should be done in this PR? I thought that that can be done separately in a follow-up PR (which I meant to track as Todo 2 of #285).

aumetra commented:
Oh, sorry, yeah makes sense! I'll remove the issue association with this PR then.

Yeah, let's keep these two things in separate PRs

aumetra commented:
Okay, so, I fucked up a PC upgrade and ordered wrong components. Now I don't have a computer right now.

Gonna review it when I manage to get it running again

aumetra reviewed (approved):
Thanks for the PR! This looks really good!
